Integrations/Digital_Defense_FrontlineVM/Digital_Defense_FrontlineVM.py

```diff
@@ -0,0 +1,795 @@
+import demistomock as demisto
+from CommonServerPython import *
+import requests
+import json
+import socket
+import struct
+import signal
+import math
+import re
+from datetime import datetime, tzinfo, timedelta
+from typing import List, Dict, Any    # throwing error: "ImportError: No module named typing"
+
+# Params:
+API_TOKEN = demisto.params().get('apiToken')
+INCIDENT_VULN_MIN_SEVERITY = demisto.params().get('incidentSeverity')
+INCIDENT_FREQUENCY = demisto.params().get('incidentFrequency')
+
+
+def get_base_url():
+    ''' Removes forward slash from end of input '''
+    url = demisto.params().get('frontlineURL')
+    url = re.sub(r'\/$', '', url)
+    return url
+
+
+# Endpoints:
+BASE_URL = get_base_url()
+VULN_ENDPOINT = BASE_URL + "/api/scanresults/active/vulnerabilities/"
+HOST_ENDPOINT = BASE_URL + "/api/scanresults/active/hosts/"
+SCAN_ENDPOINT = BASE_URL + "/api/scans/"
+
+# FrontlineVM (FVM) request header for API calls:
+FVM_HEADER = {'Authorization': 'Token ' + str(API_TOKEN)}
+
+# Minimum time to timeout functions (5 mins)
+MIN_TIMEOUT = 300   # seconds
+
+# HEADERS:
+VULN_DATA_HEADERS = ['vuln-id',
+                     'hostname',
+                     'ip-address',
+                     'vuln-title',
+                     'date-created',
+                     'ddi-severity',
+                     'vuln-info']
+
+HOST_HEADERS = ['ID',
+                'Hostname',
+                'IP',
+                'DNSHostname',
+                'MAC',
+                'OS',
+                'OSType',
+                'CriticalVulnCount']
+
+
+'''HELPER FUNCTIONS'''
+
+
+class EndOfTime(Exception):
+    ''' Raised when functions timeout '''
+    pass
+
+
+def function_timeout(signum, frame):
+    '''
+    Used to raise EndOfTime exception for timeout functions.
+    '''
+    raise EndOfTime('Function has timed out')
+
+
+def get_function_timeout_time(data_count):
+    '''
+    Returns time (in seconds) to timeout function
+    based upon the amount of data to pull.
+    '''
+    timeout_time = math.ceil(data_count / 2)
+    if timeout_time < MIN_TIMEOUT:
+        timeout_time = MIN_TIMEOUT
+    return timeout_time
```

Comment:
Why can't we replace all of this code by simply using the requests 'timeout' parameter?

Reply:
We have implemented the use of requests timeout parameter, but it was previously suggested that we timeout the `get_fvm_data` function that makes our API calls.
We assumed this was to catch being in the while loop of the function indefinitely.

So we moved the while loop in it's own function `get_all_data()` and the above code is used to setup a singal to timeout the whole function.

We can absolutely just use the requests 'timeout' parameter if you would like us to remove this code.?